runtime/remote_config.go

```diff
@@ -0,0 +1,300 @@
+	GetInt(string, int64) (int64, error)
+	GetString(string, string) (string, error)
+	GetStruct(string, interface{}) error
+	Subscribe(string, *func())
+	Unsubscribe(string)
+	Refresh() error
+	Close()
+}
+
+// RemoteConfigOptions configs remote config file path and refresh frequency
+type RemoteConfigOptions struct {
+	FilePath        string
+	PollingInterval time.Duration
+}
+
+// RemoteConfigValue represents a json serialized string
+type RemoteConfigValue struct {
+	bytes    []byte
+	dataType jsonparser.ValueType
+}
+
+// RemoteConfigMap allows lock-free concurrent rw
+// TODO consider https://golang.org/pkg/sync/#Map golang@1.9^
+type RemoteConfigMap map[string]*RemoteConfigValue
+
+type remoteConfig struct {
+	config      *RemoteConfigOptions
+	subscribers map[string]*func()
+	props       atomic.Value
+	close       chan struct{}
+	currStat    os.FileInfo
+	logger      *zap.Logger
+	scope       tally.Scope
+	poller      *time.Ticker
+	mutex       sync.RWMutex
+	wg          sync.WaitGroup
+}
+
+// NewRemoteConfig allocates a config that can be dynamically changed during runtime.
+// RemoteConfigOptions takes two args, config file path (that remote config will read from)
+// and polling interval (decides how frequent we check ).
+// The remote config file follows the same requirements with runtime/static_config/StaticConfig
+// Why RemoteConfig:
+// 		What diverge staticConfig with remoteConfig is that staticConfig allows value setting
+//		while remoteConfig is read only.
+func NewRemoteConfig(cfg *RemoteConfigOptions, logger *zap.Logger, scope tally.Scope) (RemoteConfig, error) {
+	if logger == nil {
+		logger = zap.NewNop()
+	}
+	if len(cfg.FilePath) < 1 || cfg.PollingInterval == 0 {
+		return nil, errors.Errorf("invalid remote config options")
+	}
+	if _, err := os.Stat(cfg.FilePath); os.IsNotExist(err) {
+		return nil, errors.Errorf("invalid remote config file path")
+	}
+	if scope == nil {
+		logger.Warn("no valid metrics scope")
+		scope = tally.NoopScope
+	}
+	rc := &remoteConfig{
+		config:      cfg,
+		close:       make(chan struct{}, 1),
+		subscribers: make(map[string]*func()),
+		logger:      logger,
+		scope:       scope,
+		poller:      time.NewTicker(cfg.PollingInterval),
+	}
+	rc.props.Store(make(RemoteConfigMap))
+	if err := rc.Refresh(); err != nil {
+		return nil, err
+	}
+	rc.wg.Add(1)
+	go rc.poll()
+	return rc, nil
+}
+
+func (rc *remoteConfig) loadConfig() RemoteConfigMap {
+	return rc.props.Load().(RemoteConfigMap)
+}
+
+func (rc *remoteConfig) getValidatedValue(key string, vt jsonparser.ValueType) (*RemoteConfigValue, error) {
+	ret, ok := rc.loadConfig()[key]
+	if !ok {
+		return nil, errors.Errorf("Key <%s> does not exist", key)
+	}
+	if ret.dataType != vt {
+		return nil, errors.Errorf("Key <%s> type mismatch (expected %s actual %s)", key, vt, ret.dataType)
+	}
+	return ret, nil
+}
+
+// GetBoolean returns the value as a boolean
+func (rc *remoteConfig) GetBoolean(key string) (bool, error) {
+	var (
```

Comment:
no need for this `var` statement ? can just use 

```
ret, err := rc.getValidatedValue(key, jsonparser.Boolean)
if err != nil {
  return false, err
}
v, _ := jsonparser.ParseBoolean(ret.bytes)
return v, nil
```

Reply:
good catch